# 1. Load and Explore Training Dataset

In [1]:
# Launch magic commands to automatically reload modules
%load_ext autoreload
%autoreload 2

In [2]:
# Import pandas and numpy packages
import pandas as pd
import numpy as np

## 1a. Training Dataset

In [3]:
# Load the training dataset and save as df_train
df_train = pd.read_csv('../data/raw/advmla-2023-spring/train.csv')

C:\Users\this\AppData\Local\Temp\ipykernel_9940\125755527.py:2: DtypeWarning: Columns (26) have mixed types. Specify dtype option on import or set low_memory=False.
  df_train = pd.read_csv('../data/raw/advmla-2023-spring/train.csv')


In [4]:
# Display the first 5 row of df_train
df_train.head()

team conf  GP  Min_per   Ortg   usg   eFG  TS_per  ORB_per   
0  South Alabama   SB  26     29.5   97.3  16.6  42.5   44.43      1.6  \
1       Utah St.  WAC  34     60.9  108.3  14.9  52.4   54.48      3.8   
2  South Florida   BE  27     72.0   96.2  21.8  45.7   47.98      2.1   
3     Pepperdine  WCC  30     44.5   97.7  16.0  53.6   53.69      4.1   
4        Pacific   BW  33     56.2   96.5  22.0  52.8   54.31      8.3   

   DRB_per  ...     dgbpm    oreb    dreb    treb     ast     stl     blk   
0      4.6  ... -1.941150  0.1923  0.6154  0.8077  1.1923  0.3462  0.0385  \
1      6.3  ... -0.247934  0.6765  1.2647  1.9412  1.8235  0.4118  0.2353   
2      8.0  ... -0.883163  0.6296  2.3333  2.9630  1.9630  0.4815  0.0000   
3      9.4  ... -0.393459  0.7000  1.4333  2.1333  1.1000  0.5667  0.1333   
4     18.6  ... -0.668318  1.4242  3.3030  4.7273  0.8485  0.4545  0.3333   

       pts                             player_id  drafted  
0   3.8846  7be2aead-da4e-4d13-a74b-4c1e692e2368      0.0  
1   5.9412  61de55d9-1582-4ea4-b593-44f6aa6524a6      0.0  
2  12.1852  efdc4cfc-9dd0-4bf8-acef-7273e4d5b655      0.0  
3   4.9333  14f05660-bb3c-4868-b3dd-09bcdb64279d      0.0  
4   7.5758  a58db52f-fbba-4e7b-83d0-371efcfed039      0.0  

[5 rows x 64 columns]

In [5]:
# Display the shape of df_train
df_train.shape

(56091, 64)

From the shape of dataframe df_train, we can know that this dataframe contains 64 columns and 56,091 rows.

In [6]:
# Display the information of df_train
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56091 entries, 0 to 56090
Data columns (total 64 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   team                 56091 non-null  object 
 1   conf                 56091 non-null  object 
 2   GP                   56091 non-null  int64  
 3   Min_per              56091 non-null  float64
 4   Ortg                 56091 non-null  float64
 5   usg                  56091 non-null  float64
 6   eFG                  56091 non-null  float64
 7   TS_per               56091 non-null  float64
 8   ORB_per              56091 non-null  float64
 9   DRB_per              56091 non-null  float64
 10  AST_per              56091 non-null  float64
 11  TO_per               56091 non-null  float64
 12  FTM                  56091 non-null  int64  
 13  FTA                  56091 non-null  int64  
 14  FT_per               56091 non-null  float64
 15  twoPM                56091 non-null 

In [7]:
# Display the descriptive statistics of df_train
df_train.describe()

GP       Min_per          Ortg           usg           eFG   
count  56091.000000  56091.000000  56091.000000  56091.000000  56091.000000  \
mean      23.411973     37.325229     91.858295     18.149265     44.575513   
std       10.090542     28.061897     30.538819      6.204489     18.213813   
min        1.000000      0.000000      0.000000      0.000000      0.000000   
25%       16.000000      9.500000     84.100000     14.500000     40.100000   
50%       28.000000     35.900000     97.100000     18.100000     47.700000   
75%       31.000000     62.200000    106.900000     21.800000     53.200000   
max       41.000000     98.000000    834.000000     50.000000    150.000000   

             TS_per       ORB_per       DRB_per       AST_per        TO_per   
count  56091.000000  56091.000000  56091.000000  56091.000000  56091.000000  \
mean      47.676882      5.575376     12.673716     10.848740     20.337819   
std       17.414133      9.480239     10.878099      9.356001     12.226129   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%       43.930000      1.800000      8.400000      4.600000     14.800000   
50%       50.850000      4.300000     11.900000      9.100000     19.100000   
75%       56.060000      8.100000     16.000000     15.100000     24.500000   
max      150.000000   1576.600000   1385.000000    100.000000    100.000000   

       ...         ogbpm         dgbpm          oreb          dreb   
count  ...  56047.000000  56047.000000  56053.000000  56053.000000  \
mean   ...     -2.083375     -0.442297      0.770611      1.892593   
std    ...      6.190166      3.279438      0.738899      1.479148   
min    ...   -141.539000   -100.984000      0.000000      0.000000   
25%    ...     -4.426515     -1.728030      0.225800      0.700000   
50%    ...     -1.580380     -0.313051      0.545500      1.633300   
75%    ...      0.992335      1.066415      1.111100      2.758600   
max    ...    225.485000     78.985500      5.933300     11.545500   

               treb           ast           stl           blk           pts   
count  56053.000000  56053.000000  56053.000000  56053.000000  56053.000000  \
mean       2.663204      1.073038      0.529475      0.282825      5.775381   
std        2.099913      1.171148      0.469152      0.417754      4.953723   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        1.000000      0.222200      0.160000      0.000000      1.571400   
50%        2.259300      0.677400      0.424200      0.133300      4.482800   
75%        3.833300      1.533300      0.793100      0.366700      9.069000   
max       14.533300     10.000000      4.000000      5.258100     30.090900   

            drafted  
count  56091.000000  
mean       0.009556  
std        0.097287  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000  

[8 rows x 57 columns]

## 1b. Testing Dataset

In [8]:
# Load the training dataset and save as df_test
df_test = pd.read_csv('../data/raw/advmla-2023-spring/test.csv')

In [9]:
# Display the first 5 row of df_test
df_test.head()

team  conf  GP  Min_per   Ortg   usg   eFG  TS_per  ORB_per   
0          Morgan St.  MEAC   2      3.0  115.1   4.7  50.0   50.00      0.0  \
1  South Carolina St.  MEAC  11     17.6   61.1  18.6  34.7   35.18      2.5   
2          Binghamton    AE   9     28.6   91.9  23.8  54.1   52.49      6.4   
3            Illinois   B10   7      1.3  111.0  10.4  83.3   83.33      0.0   
4            Iowa St.   B12  23     78.5  103.1  21.5  54.0   56.12      3.6   

   DRB_per  ...     ogbpm     dgbpm    oreb    dreb    treb     ast     stl   
0      4.6  ... -2.467740 -2.275660  0.0000  0.3333  0.3333  0.0000  0.0000  \
1     15.7  ... -7.494720 -4.412530  0.2727  1.4545  1.7273  0.4545  0.1818   
2     22.5  ... -2.924950  1.717890  1.3333  4.4444  5.7778  1.0000  0.6667   
3     13.4  ... -0.767911  0.962469  0.0000  0.2857  0.2857  0.0000  0.0000   
4     10.2  ...  2.893610 -1.019000  1.0435  2.8696  3.9130  1.1739  0.8261   

      blk      pts                             player_id  
0  0.0000   1.0000  cf302b4d-84f7-4124-a25d-a75eed31978b  
1  0.0000   2.3636  f91837cd-4f49-4b70-963d-aeb82c6ce3da  
2  1.8889   8.8889  53ec2a29-1e7d-4c6d-86d7-d60d02af8916  
3  0.0000   0.7143  32402798-471c-4a54-8cb4-29cd95199014  
4  0.0870  14.3043  73b960f9-27b8-4431-9d23-a760e9bbc360  

[5 rows x 63 columns]

In [10]:
# Display the shape of df_test
df_test.shape

(4970, 63)

From the shape of dataframe df_test, we can know that this dataframe contains 63 columns and 4,970 rows.

In [11]:
# Display the information of df_test
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4970 entries, 0 to 4969
Data columns (total 63 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   team                 4970 non-null   object 
 1   conf                 4970 non-null   object 
 2   GP                   4970 non-null   int64  
 3   Min_per              4970 non-null   float64
 4   Ortg                 4970 non-null   float64
 5   usg                  4970 non-null   float64
 6   eFG                  4970 non-null   float64
 7   TS_per               4970 non-null   float64
 8   ORB_per              4970 non-null   float64
 9   DRB_per              4970 non-null   float64
 10  AST_per              4970 non-null   float64
 11  TO_per               4970 non-null   float64
 12  FTM                  4970 non-null   int64  
 13  FTA                  4970 non-null   int64  
 14  FT_per               4970 non-null   float64
 15  twoPM                4970 non-null   i

In [12]:
# Display the descriptive statistics of df_test
df_test.describe()

GP      Min_per         Ortg          usg          eFG   
count  4970.000000  4970.000000  4970.000000  4970.000000  4970.000000  \
mean     15.865795    34.906881    89.500644    17.867626    43.657203   
std       8.276655    27.921356    35.338827     6.780159    20.716630   
min       1.000000     0.100000     0.000000     0.000000     0.000000   
25%       9.000000     6.700000    80.900000    14.200000    38.400000   
50%      17.000000    32.350000    96.800000    18.000000    48.100000   
75%      22.000000    59.900000   107.100000    21.800000    54.000000   
max      33.000000    96.000000   391.200000    48.200000   150.000000   

            TS_per      ORB_per      DRB_per      AST_per       TO_per  ...   
count  4970.000000  4970.000000  4970.000000  4970.000000  4970.000000  ...  \
mean     46.538342     5.168390    13.048753    10.356801    18.962254  ...   
std      19.992005     7.030561     9.248595     9.427192    13.255566  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%      42.340000     1.400000     8.400000     4.300000    13.100000  ...   
50%      50.850000     3.700000    12.300000     8.600000    18.100000  ...   
75%      56.780000     7.400000    16.800000    14.500000    23.400000  ...   
max     150.000000   231.800000   121.100000   100.000000   100.000000  ...   

                mp        ogbpm        dgbpm         oreb         dreb   
count  4970.000000  4969.000000  4969.000000  4970.000000  4970.000000  \
mean     16.569559    -2.149284    -0.526065     0.697025     1.953592   
std      10.550727     6.750792     3.920493     0.683636     1.533018   
min       0.250000   -62.101300   -31.594500     0.000000     0.000000   
25%       6.891675    -4.443960    -1.868180     0.200000     0.750000   
50%      16.083350    -1.540360    -0.397400     0.500000     1.666700   
75%      25.991375     1.103500     1.049540     1.000000     2.882400   
max      38.800000   160.805000    71.811500     5.148100    10.681800   

              treb          ast          stl          blk          pts  
count  4970.000000  4970.000000  4970.000000  4970.000000  4970.000000  
mean      2.650609     1.069840     0.529118     0.260375     5.753262  
std       2.073589     1.153074     0.474838     0.380465     4.881835  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%       1.000000     0.250000     0.142900     0.000000     1.602175  
50%       2.260900     0.680000     0.428600     0.133300     4.535900  
75%       3.800000     1.533300     0.807700     0.333300     9.000000  
max      15.000000     8.400000     3.565200     3.640000    25.333300  

[8 rows x 57 columns]

# 2. Data Preparation

## 2a. Training Dataset

In [13]:
# Make a copy of df_train and save as df_train_cleaned
df_train_cleaned = df_train.copy()

First, we are going to remove columns whose datatype is object.

In [15]:
# Remove columns with objects datatype from the dataset
df_train_cleaned.drop(df_train[['team', 'conf', 'yr', 'ht', 'num', 'type', 'player_id']], axis=1, inplace=True)

In [16]:
# Check if there is any columns with missing values
df_train_cleaned.isna().sum()

GP                         0
Min_per                    0
Ortg                       0
usg                        0
eFG                        0
TS_per                     0
ORB_per                    0
DRB_per                    0
AST_per                    0
TO_per                     0
FTM                        0
FTA                        0
FT_per                     0
twoPM                      0
twoPA                      0
twoP_per                   0
TPM                        0
TPA                        0
TP_per                     0
blk_per                    0
stl_per                    0
ftr                        0
porpag                     0
adjoe                      0
pfr                        0
year                       0
Rec_Rank               39055
ast_tov                 4190
rimmade                 6081
rimmade_rimmiss         6081
midmade                 6081
midmade_midmiss         6081
rim_ratio               9464
mid_ratio               9688
dunksmade     

For those columns with missing values more than 50% of the total values, we decide to remove them from the dataframe. For the rest, we will fill the missing values with the mean value of that column. 

In [17]:
# Remove columns with more than 50% missing values from the dataframe
df_train_cleaned.drop(df_train[['Rec_Rank', 'dunks_ratio', 'pick']], axis=1, inplace=True)

In [18]:
# Replace all missing values with mean values of corresponding columns
df_train_cleaned = df_train_cleaned.fillna(df_train_cleaned.mean())

In [19]:
# Check if there is still any missing values
df_train_cleaned.isna().sum()

GP                     0
Min_per                0
Ortg                   0
usg                    0
eFG                    0
TS_per                 0
ORB_per                0
DRB_per                0
AST_per                0
TO_per                 0
FTM                    0
FTA                    0
FT_per                 0
twoPM                  0
twoPA                  0
twoP_per               0
TPM                    0
TPA                    0
TP_per                 0
blk_per                0
stl_per                0
ftr                    0
porpag                 0
adjoe                  0
pfr                    0
year                   0
ast_tov                0
rimmade                0
rimmade_rimmiss        0
midmade                0
midmade_midmiss        0
rim_ratio              0
mid_ratio              0
dunksmade              0
dunksmiss_dunksmade    0
drtg                   0
adrtg                  0
dporpag                0
stops                  0
bpm                    0


So all the rows have been filled with values. We will take a look at the cleaned dataframe before pre-processing it.

In [20]:
# Show the first 5 rows of the df_train_cleaned dataframe
df_train_cleaned.head()

GP  Min_per   Ortg   usg   eFG  TS_per  ORB_per  DRB_per  AST_per  TO_per   
0  26     29.5   97.3  16.6  42.5   44.43      1.6      4.6     15.8    16.3  \
1  34     60.9  108.3  14.9  52.4   54.48      3.8      6.3     13.6    19.8   
2  27     72.0   96.2  21.8  45.7   47.98      2.1      8.0     14.7    15.9   
3  30     44.5   97.7  16.0  53.6   53.69      4.1      9.4     13.7    23.8   
4  33     56.2   96.5  22.0  52.8   54.31      8.3     18.6      8.2    22.7   

   ...     ogbpm     dgbpm    oreb    dreb    treb     ast     stl     blk   
0  ... -2.781990 -1.941150  0.1923  0.6154  0.8077  1.1923  0.3462  0.0385  \
1  ... -0.052263 -0.247934  0.6765  1.2647  1.9412  1.8235  0.4118  0.2353   
2  ...  1.548230 -0.883163  0.6296  2.3333  2.9630  1.9630  0.4815  0.0000   
3  ... -0.342775 -0.393459  0.7000  1.4333  2.1333  1.1000  0.5667  0.1333   
4  ... -1.684860 -0.668318  1.4242  3.3030  4.7273  0.8485  0.4545  0.3333   

       pts  drafted  
0   3.8846      0.0  
1   5.9412      0.0  
2  12.1852      0.0  
3   4.9333      0.0  
4   7.5758      0.0  

[5 rows x 54 columns]

## 2b. Testing Dataset

In [21]:
# Make a copy of df_test and save as df_test_cleaned
df_test_cleaned = df_test.copy()

First, we are going to remove columns whose datatype is object.

In [23]:
# Remove columns with objects datatype from the dataset
df_test_cleaned.drop(df_test[['team', 'conf', 'yr', 'ht', 'num', 'type', 'player_id']], axis=1, inplace=True)

In [24]:
# Check if there is any columns with missing values
df_test_cleaned.isna().sum()

GP                        0
Min_per                   0
Ortg                      0
usg                       0
eFG                       0
TS_per                    0
ORB_per                   0
DRB_per                   0
AST_per                   0
TO_per                    0
FTM                       0
FTA                       0
FT_per                    0
twoPM                     0
twoPA                     0
twoP_per                  0
TPM                       0
TPA                       0
TP_per                    0
blk_per                   0
stl_per                   0
ftr                       0
porpag                    0
adjoe                     0
pfr                       0
year                      0
Rec_Rank               3536
ast_tov                 537
rimmade                 248
rimmade_rimmiss         248
midmade                 248
midmade_midmiss         248
rim_ratio               646
mid_ratio               697
dunksmade               248
dunksmiss_dunksmade 

For those columns with missing values more than 50% of the total values, we decide to remove them from the dataframe. For the rest, we will fill the missing values with the mean value of that column.

In [25]:
# Remove columns with more than 50% missing values from the dataframe
df_test_cleaned.drop(df_test[['Rec_Rank', 'dunks_ratio', 'pick']], axis=1, inplace=True)

In [26]:
# Replace all missing values with mean values of corresponding columns
df_test_cleaned = df_test_cleaned.fillna(df_test_cleaned.mean())

In [27]:
# Check if there is still any missing values
df_test_cleaned.isna().sum()

GP                     0
Min_per                0
Ortg                   0
usg                    0
eFG                    0
TS_per                 0
ORB_per                0
DRB_per                0
AST_per                0
TO_per                 0
FTM                    0
FTA                    0
FT_per                 0
twoPM                  0
twoPA                  0
twoP_per               0
TPM                    0
TPA                    0
TP_per                 0
blk_per                0
stl_per                0
ftr                    0
porpag                 0
adjoe                  0
pfr                    0
year                   0
ast_tov                0
rimmade                0
rimmade_rimmiss        0
midmade                0
midmade_midmiss        0
rim_ratio              0
mid_ratio              0
dunksmade              0
dunksmiss_dunksmade    0
drtg                   0
adrtg                  0
dporpag                0
stops                  0
bpm                    0


So all the rows have been filled with values. We will take a look at the cleaned dataframe before pre-processing it.

In [28]:
# Show the first 5 rows of the df_test_cleaned dataframe
df_test_cleaned.head()

GP  Min_per   Ortg   usg   eFG  TS_per  ORB_per  DRB_per  AST_per  TO_per   
0   2      3.0  115.1   4.7  50.0   50.00      0.0      4.6      0.0     0.0  \
1  11     17.6   61.1  18.6  34.7   35.18      2.5     15.7      8.1    30.4   
2   9     28.6   91.9  23.8  54.1   52.49      6.4     22.5     10.0    21.9   
3   7      1.3  111.0  10.4  83.3   83.33      0.0     13.4      0.0    33.6   
4  23     78.5  103.1  21.5  54.0   56.12      3.6     10.2      7.8    16.1   

   ...       mp     ogbpm     dgbpm    oreb    dreb    treb     ast     stl   
0  ...  10.0000 -2.467740 -2.275660  0.0000  0.3333  0.3333  0.0000  0.0000  \
1  ...  11.5455 -7.494720 -4.412530  0.2727  1.4545  1.7273  0.4545  0.1818   
2  ...  23.3333 -2.924950  1.717890  1.3333  4.4444  5.7778  1.0000  0.6667   
3  ...   2.2857 -0.767911  0.962469  0.0000  0.2857  0.2857  0.0000  0.0000   
4  ...  32.7826  2.893610 -1.019000  1.0435  2.8696  3.9130  1.1739  0.8261   

      blk      pts  
0  0.0000   1.0000  
1  0.0000   2.3636  
2  1.8889   8.8889  
3  0.0000   0.7143  
4  0.0870  14.3043  

[5 rows x 53 columns]

## 2c. Set features and target variables

In [29]:
# Drop column 'drafted' from df_train_cleaned and save as an variable called X_data
X_data = df_train_cleaned.drop('drafted', axis='columns')

In [30]:
# Save name of the target column as a variable called y_data
y_data = df_train_cleaned['drafted']

In [31]:
# Print the shape of X_data and y_data
print(X_data.shape)
print(y_data.shape)

(56091, 53)
(56091,)


In [32]:
# Assign df_test_cleaned as an variable called X_test
X_test = df_test_cleaned

# 3. Split Dataset

In [34]:
# Import the train_test_split package from scikit-learn
from sklearn.model_selection import train_test_split

In [35]:
# Split the training dataset to training data (80%) and validation data (20%)
X_train, X_val, y_train, y_val = train_test_split(X_data, y_data, test_size=0.2, random_state=42, stratify=y_data)

In [36]:
# Print the shape of X_train, X_val and X_test
print(X_train.shape)
print(X_val.shape)
print(X_test.shape)

(44872, 53)
(11219, 53)
(4970, 53)


In [37]:
# Print the shape of y_train, y_val
print(y_train.shape)
print(y_val.shape)

(44872,)
(11219,)


# 4. Scale Data

In [38]:
# Import StandardScaler from scikit-learn to scale the data
from sklearn.preprocessing import StandardScaler

In [39]:
# Instantiate the StandardScaler method and save as a variable called scaler
scaler = StandardScaler()

In [40]:
# Fit the training dataset to the scaler
scaler.fit(X_train)

StandardScaler()

In [41]:
# Apply the scaler to the training, validation and testing datasets
X_train = scaler.transform(X_train)
X_val = scaler.transform(X_val)
X_test = scaler.transform(X_test)

In [42]:
# Import dump from joblib
from joblib import dump

In [43]:
# Save scaler into the folder models and call the file as scaler.joblib
dump(scaler, '../models/scaler.joblib')

['../models/scaler.joblib']

In [45]:
# Save the sets to folder data/processed
pd.DataFrame(X_train).to_csv('../data/processed/X_train.csv', index=False)
pd.DataFrame(X_val).to_csv('../data/processed/X_val.csv', index=False)
pd.DataFrame(X_test).to_csv('../data/processed/X_test.csv', index=False)
pd.DataFrame(y_train).to_csv('../data/processed/y_train.csv', index=False)
pd.DataFrame(y_val).to_csv('../data/processed/y_val.csv', index=False)

# 5. Assess Baseline Model

In [46]:
# Find the value which is most present in the target variable
y_mode = y_train.mode()
y_mode

0    0.0
Name: drafted, dtype: float64

In [47]:
# Create a numpy array with the same dimensions as for the training set called y_base filled with this value
y_base = np.full(y_train.shape, y_mode)